pkg/storage/formattedstore/batchwrite.go

```diff
@@ -27,78 +27,83 @@ type batchStore interface {
 
 // BatchService is a batch service.
 type BatchService struct {
-	sizeLimit int
-	timeLimit time.Duration
 	sync.RWMutex
-	values       models.Batch
-	futureValues map[string]*batch
-	formatter    Formatter
-	provider     batchProvider // TODO remove batch provider after refactoring edv batch
-	addBatchChan chan *batch
+	values           models.Batch
+	futureValues     map[string]*batch
+	futureValuesLock sync.RWMutex
+	formatter        Formatter
+	provider         batchProvider // TODO remove batch provider after refactoring edv batch
+	addBatchChan     chan *batch
+	readyChan        chan struct{}
 }
 
 type batch struct {
 	keyID                   string
 	value                   []byte
 	isDeleted               bool
 	addEncryptedIndicesFunc func(k string, encryptedDocumentBytes []byte) (*models.EncryptedDocument, error)
+	isReady                 bool
 }
 
 // NewBatchWrite new batch write.
-func NewBatchWrite(sizeLimit int, timeLimit time.Duration, formatter Formatter,
-	provider batchProvider) *BatchService {
-	ticker := time.NewTicker(timeLimit)
-	quit := make(chan struct{})
+//nolint: funlen
+func NewBatchWrite(batchThreadLimit int, formatter Formatter, provider batchProvider) *BatchService {
 	b := &BatchService{
-		sizeLimit: sizeLimit, timeLimit: timeLimit, values: make(models.Batch, 0),
-		futureValues: make(map[string]*batch), formatter: formatter, provider: provider,
-		addBatchChan: make(chan *batch, sizeLimit),
+		values: make(models.Batch, 0), futureValues: make(map[string]*batch), formatter: formatter, provider: provider,
+		addBatchChan: make(chan *batch, batchThreadLimit), readyChan: make(chan struct{}),
 	}
 
 	go func(b *BatchService) {
-		for {
-			select {
-			case <-ticker.C:
-				if err := b.flush(); err != nil {
-					logger.Errorf(err.Error())
+		for addBatch := range b.addBatchChan {
+			if addBatch.isReady {
+				b.readyChan <- struct{}{}
+				return
+			}
+
+			if addBatch.isDeleted {
+				b.Lock()
+
+				id, err := b.formatter.GenerateEDVCompatibleID(addBatch.keyID)
```

Comment:
If this batch service/formatter is specific to edv, then better to move this to /storage/edv rather than here (/storage/formattedstore) and inject the batch service (probably a future PR).

Reply:
this will be part of refactoring storage interface

Reply:
storage interface need flush method

Reply:
will be done in v0.1.6